### PR TITLE
Rewrite test/models in TypeScript

### LINKED ISSUE #7864: Rewrite bokehjs' tests in TypeScript
A follow up on #6481. Writing tests in TypeScript is an important part of verifying that our typings actually work, especially parts that can't be verified within the lib itself. This is also an opportunity to clean up the atrocious mess that bokehjs' unit tests are.

Rewrite progress:

- [x] client
- [x] core
- [x] document
- [x] models
- [x] protocol

### PR DESCRIPTION
**Pull request by mattpap**
After this PR `test/index`, `test/document/document`, `test/core/ui_events` and `test/client/connection` will still need a rewrite.

addresses #7864

### CODE REVIEW

File: bokehjs/test/models/mappers/categorical_color_mapper.ts [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+            end: j,
+          })
+
+          const vals0 = cm0.v_compute(["a"])
+          expect(vals0).to.be.deep.equal(["gray"])
+
+          /* XXX: gives ["red"] because original code used [i, j] = [3, 4] due to lack of lexical scoping
+          const vals1 = cm0.v_compute([["a", "1"]])
+          expect(vals1).to.be.deep.equal(["gray"])
+          */
```
Comment (mattpap) [inline bokehjs/test/models/mappers/categorical_color_mapper.ts]: @bryevdv, the original code always used `[i, j] = [3, 4]` due to CoffeeScript's messed up scoping rules (`for` loop without `do` statement). This now fails with `["red"]` result, but I thing this is actually correct behavior.

File: bokehjs/test/models/mappers/categorical_marker_mapper.ts [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+            markers,
+            factors: factors.map((x) => x.slice(0, 2) as F2),
+            start: i,
+            end: j,
+          })
+
+          const vals0 = cm0.v_compute(["a"])
+          expect(vals0).to.be.deep.equal(["circle"])
+
+          /* XXX: gives ["hex"] because original code used [i, j] = [3, 4] due to lack of lexical scoping
```
Comment (mattpap) [inline bokehjs/test/models/mappers/categorical_marker_mapper.ts]: @bryevdv, similarly here, just fails with `["hex"]`, which I think is correct.

File: bokehjs/test/protocol/message.ts [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        expect(m.buffers).to.be.deep.equal([])
+      })
+    })
+
+    describe("assemble_buffer method", () => {
+      const m = Message.create("FOO", {bar:2}, {baz:3})
+      m.header.num_buffers = 2
+
+      it("should append a new buffer", () => {
+        m.assemble_buffer(1, 2)
```
Comment (mattpap) [inline bokehjs/test/protocol/message.ts]: @bryevdv, I suppose that arguments to `assemble_buffer()` are just whatever/throwaway values (they don't match types I figured out in `src/lib/protocol/message.ts`)?

File: bokehjs/test/models/mappers/categorical_color_mapper.ts (hunk 1 above)
Comment (bryevdv) [reply bokehjs/test/models/mappers/categorical_color_mapper.ts]: I concur it should be `"red"`

File: bokehjs/test/models/mappers/categorical_marker_mapper.ts (hunk 2 above)
Comment (bryevdv) [reply bokehjs/test/models/mappers/categorical_marker_mapper.ts]: Yes, i think this specific one should actually map to `hex` as correct behavior

File: bokehjs/test/protocol/message.ts (hunk 3 above)
Comment (bryevdv) [reply bokehjs/test/protocol/message.ts]: the values here in the test are just junk, yes

**Pull request merged**